Proper behavior of Eurotronic SPZB0001 that doesn't support presetmodes or have heat as boost mode in deCONZ integration

KartoffelToby opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
nothing

## Proposed change
In a nutshell _Add a setting to deconz intigration for remap modes to match the device_

As an example Euronics Spirit has no preset modes, and mode heat is boost, auto means controlled by the target temperature and off is off. So auto is the default mode on this devices and the devices self has two types of modes Heat and Off.
DECONZ | off -> No heating -> off | HA
DECONZ | auto -> Heating to defined target temperature -> heat | HA
DECONZ | heat -> Set target to 30°C boost heat -> boost | HA

It should be possible to tell HA to interpret the auto mode as heating mode. This is not only for HA, if GA Intigration is used, it's not possible to controll the target temperature with Google Assistient as the mode is set to auto and auto in google means controlled by AI, but the Spirit still uses a target set temperature in "deconz auto mode".

I also added a min and max temp setting for GA because deconz uses 5°C - 30°C and GA default 10°C - 32°C

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
- Add a new setting to deconz intigration called "set_auto_to_heat_ga" to switch between default behavior and remapping mode
- Add min and max temp to GA intigration
- Fix modes in GA intigration, GA require a array. A comma seperated list is deprecated. 

- This PR fixes or closes issue: fixes #47768

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

chemelli74 commented:
Hi @KartoffelToby , I think you should rebase your PR.

Simone

KartoffelToby commented:
@chemelli74 Sure, something was messed up my fault. Now, everything should be right.

thecode commented:
@frenck can I ask you to clean the tags from this PR? It wrongly tags to much integrations, Thanks.

Kane610 commented on homeassistant/components/google_assistant/trait.py:
```diff
@@ -744,6 +744,12 @@ def sync_attributes(self):
             self.hass.config.units.temperature_unit
         )
 
+        # Add a default min max value for set temperature to display correctly in Google Assistant App
+        response["thermostatTemperatureRange"] = {
+            "minThresholdCelsius": 5.0,
+            "maxThresholdCelsius": 30.0,
+        }
+
```
This will most probably not be accepted. Anyway I think it warrants its own PR, I don't really know much about the GA integration

KartoffelToby replied:
I removed the setting and added a direct override in deconz plugin for SPZB0001 devices.

Kane610 reviewed (changes requested):
Thanks for contributing! If the device behavior is different from how its been implemented its fine to change to reflect its actual behavior, but the PR will need to change before it can be accepted. Start by removing the configurable option and try to isolate the logic to within the thermostat class.

Im not sure

Kane610 commented on homeassistant/components/deconz/climate.py:
```diff
@@ -115,7 +118,18 @@ def __init__(self, device, gateway):
-        self._hvac_mode_to_deconz = dict(HVAC_MODE_TO_DECONZ)
+        # if settings climate SPZB0001 auto is heat
+        self.auto_to_heat_mapping = False
+        temp_modes = HVAC_MODE_TO_DECONZ
+        if self._device.modelid == "SPZB0001":
+            temp_modes = {
+                HVAC_MODE_COOL: "cool",
+                HVAC_MODE_HEAT: "auto",
+                HVAC_MODE_OFF: "off",
+            }
+            self.auto_to_heat_mapping = True
+        # End Device change for SPZB0001
+        self._hvac_mode_to_deconz = dict(temp_modes)
```
```suggestion
        self._hvac_mode_to_deconz = dict(HVAC_MODE_TO_DECONZ)
        if self._device.modelid == "SPZB0001":  # description about why this is needed
            self._hvac_mode_to_deconz = {
                HVAC_MODE_COOL: "cool",
                HVAC_MODE_HEAT: "auto",
                HVAC_MODE_OFF: "off",
            }
```

Kane610 commented on homeassistant/components/deconz/climate.py:
```diff
@@ -187,7 +215,9 @@ async def async_set_hvac_mode(self, hvac_mode: str) -> None:
 
         data = {"mode": self._hvac_mode_to_deconz[hvac_mode]}
-        if len(self._hvac_mode_to_deconz) == 2:  # Only allow turn on and off thermostat
+        if (
+            len(self._hvac_mode_to_deconz) == 2 and not self.auto_to_heat_mapping
+        ):  # Only allow turn on and off thermostat
```
```suggestion
        if len(self._hvac_mode_to_deconz) == 2:  # Only allow turn on and off thermostat
```

len(self._hvac_mode_to_deconz) is 3 for SPZB0001

Kane610 commented on homeassistant/components/deconz/const.py:
```diff
@@ -23,6 +23,7 @@
 DEFAULT_ALLOW_DECONZ_GROUPS = True
 DEFAULT_ALLOW_NEW_DEVICES = True
 
+
```
```suggestion
```

Kane610 commented on homeassistant/components/deconz/climate.py:
```diff
@@ -68,6 +69,8 @@
 
+DEFAULT_MIN_TEMP = 5
+DEFAULT_MAX_TEMP = 30
 
```
It doesn't look like you have the dev environment set up correctly, else the commit hooks would have caught this.

Kane610 reviewed (changes requested):
Closer!

Remove all changes to google assistant.

Create a global for the modelid and add a motivating description to the special handling of the modelid .

Also a new test that covers the different modes for this device. Take inspiration from this PR how to structure your new test https://github.com/home-assistant/core/pull/47821/files

Kane610 commented on tests/components/deconz/test_climate.py:
```diff
@@ -549,6 +549,120 @@ async def test_climate_device_with_fan_support(hass, aioclient_mock):
         )
 
 
+async def test_climate_device_SPZB0001(hass, aioclient_mock):
```
The test doesn't follow the structure of the PR I mentioned. You might need to rebase your PR against dev to make it work.

Kane610 commented on homeassistant/components/deconz/climate.py:
```diff
@@ -115,13 +124,21 @@ def __init__(self, device, gateway):
         super().__init__(device, gateway)
 
-        self._hvac_mode_to_deconz = dict(HVAC_MODE_TO_DECONZ)
+        # if settings climate SPZB0001 auto is heat
+        if self._device.modelid == "SPZB0001":
+            self._hvac_mode_to_deconz = dict(HVAC_MODE_TO_DECONZ_SPZB0001)
+        else:
+            self._hvac_mode_to_deconz = dict(HVAC_MODE_TO_DECONZ)
+
         if "mode" not in device.raw["config"]:
```
```suggestion
        self._hvac_mode_to_deconz = dict(HVAC_MODE_TO_DECONZ)

        # if settings climate SPZB0001 auto is heat
        if self._device.modelid == EUROTRONIC_SPZB0001:
            self._hvac_mode_to_deconz = {
                HVAC_MODE_HEAT: "auto",
                HVAC_MODE_OFF: "off",
            }
        elif "mode" not in device.raw["config"]:
```

Kane610 commented on homeassistant/components/deconz/climate.py:
```diff
@@ -115,13 +124,21 @@ def __init__(self, device, gateway):
                 HVAC_MODE_OFF: False,
             }
-        elif "coolsetpoint" not in device.raw["config"]:
+        elif (
+            "coolsetpoint" not in device.raw["config"]
+            and self._device.modelid != "SPZB0001"
+        ):
```
```suggestion
        elif "coolsetpoint" not in device.raw["config"]:
```

KartoffelToby replied:
We need this because the Spirit has only two modes and no coolsetpoint so that would throw an error.

Kane610 replied:
That is managed by the if statement on line 128

Kane610 commented on homeassistant/components/deconz/climate.py:
```diff
@@ -115,13 +124,21 @@ def __init__(self, device, gateway):
         super().__init__(device, gateway)
 
-        self._hvac_mode_to_deconz = dict(HVAC_MODE_TO_DECONZ)
+        # if settings climate SPZB0001 auto is heat
```
Maybe a bit more descriptive?

Kane610 commented on homeassistant/components/deconz/climate.py:
```diff
@@ -187,7 +218,9 @@ async def async_set_hvac_mode(self, hvac_mode: str) -> None:
         data = {"mode": self._hvac_mode_to_deconz[hvac_mode]}
-        if len(self._hvac_mode_to_deconz) == 2:  # Only allow turn on and off thermostat
+        if (
+            len(self._hvac_mode_to_deconz) == 2 and self._device.modelid != "SPZB0001"
```
```suggestion
            len(self._hvac_mode_to_deconz) == 2 and self._device.modelid != EUROTRONIC_SPZB0001
```

KartoffelToby replied:
Same here SPZB0001 has only two modes but is not controlled by on off

Kane610 replied:
Yes in this case it seems to be necessary

Kane610 commented on homeassistant/components/deconz/climate.py:
```diff
@@ -51,6 +52,11 @@
     HVAC_MODE_OFF: "off",
 }
 
+HVAC_MODE_TO_DECONZ_SPZB0001 = {
+    HVAC_MODE_HEAT: "auto",
+    HVAC_MODE_OFF: "off",
+}
```
```suggestion
```

KartoffelToby replied:
Sorry for wasting your time, I misunderstood you with the global variable. now it is as expected

Kane610 commented on homeassistant/components/deconz/climate.py:
```diff
@@ -68,6 +74,9 @@
 
+DEFAULT_MIN_TEMP = 5
+DEFAULT_MAX_TEMP = 30
+
```
```suggestion

EUROTRONIC_SPZB0001 = "SPZB0001"

```

Kane610 reviewed (changes requested):
Easy stuff left

KartoffelToby committed:
[TASK] rebase

[TASK] better description

[TASK] test and last changes

[TASK] use new websocket fix in deconz climate test

[TASK] fix in deconz climate test

KartoffelToby commented:
I removed my ga changes and create a new PR #48029 for that.

Thanks for your patience @Kane610

Kane610 commented:
> I removed my ga changes and create a new PR #48029 for that.
> 
> Thanks for your patience @Kane610

No worries, I'm not the best reviewer

Kane610 reviewed (changes requested):
I think its the test and my new comments left now

Kane610 commented on tests/components/deconz/test_climate.py:
```diff
@@ -668,6 +668,97 @@ async def test_climate_device_with_preset(hass, aioclient_mock):
         )
 
 
+async def test_climate_device_with_euronics(hass, aioclient_mock):
```
This test needs to be updated to follow the structure as is implemented in this PR https://github.com/home-assistant/core/pull/47821/files.

1. Don't use deepcopy but rather patch.dict
2. Don't use gateway.api.event_handler but rather the new websocket fixture

Kane610 commented on homeassistant/components/deconz/climate.py:
```diff
@@ -140,6 +157,26 @@ def supported_features(self):
+    def min_temp(self) -> float:
+        """Return the minimum temperature.
+
+        At this moment deConz doesn't return a min temp
```
```suggestion
        At this moment deCONZ doesn't return a min temp.
```

Kane610 replied:
This is not fixed yet

Kane610 commented on homeassistant/components/deconz/climate.py:
```diff
@@ -116,7 +122,18 @@ def __init__(self, device, gateway):
-        if "mode" not in device.raw["config"]:
+        """Handle Euronics Spirit Zigbee
+
+        if device is SPZB0001 its a Euronics Spirit Zigbee Thermostat,
+        it has mode heat for Boost and auto is eq to heat.
+        so we remap the HVAC Modes to match the device in deconz
```
```suggestion
        If modelid is SPZB0001 it's a Euronics Spirit Zigbee Thermostat,
        it has mode heat for Boost and auto is eq to heat.
        So we remap the HVAC Modes to match the device in deCONZ.
```

Kane610 replied:
This is not fixed yet.

Kane610 commented on homeassistant/components/deconz/climate.py:
```diff
@@ -140,6 +157,26 @@ def supported_features(self):
+    def max_temp(self) -> float:
+        """Return the maximum temperature.
+
+        At this moment deConz doesn't return a max temp
```
```suggestion
        At this moment deCONZ doesn't return a max temp.
```

Kane610 commented on tests/components/deconz/test_climate.py:
```diff
@@ -668,6 +668,97 @@ async def test_climate_device_with_preset(hass, aioclient_mock):
 
 
+async def test_climate_device_with_euronics(hass, aioclient_mock):
+    """Test successful creation of sensor entities."""
```
Better description about what you want to validate with this test.

Kane610 reviewed (changes requested):
Looks good, fix formatting of strings and the **test**.

Kane610 reviewed (changes requested):
Don't do the whole test file, just your own test. The other PR will cover the other tests.

KartoffelToby commented:
> Don't do the whole test file, just your own test. The other PR will cover the other tests.

Yes, i'm not sure about the import, if it merged correctly i'll be fine

Kane610 commented:
> > Don't do the whole test file, just your own test. The other PR will cover the other tests.
> 
> Yes, i'm not sure about the import, if it merged correctly i'll be fine

The other PR just got merged so a rebase will solve it for you

Kane610 reviewed (approved):
Looks good! Can be merged when rebase has been done and tests pass

Kane610 commented:
Awesome

KartoffelToby commented:
@Kane610 think its ready to merge now.

Kane610 commented:
> @Kane610 think its ready to merge now.

I had a discussion with Frenck around thid item and we need to have a chat you and me if this is something even more generic than for just you're device. But I dont have time yet to do that :)

Kane610 commented:
So Frenck thought it was wrong to put a quirk like this into the integration, there are two ways forward here, this is an issue that should be fixed in deconz, or I haven't done the work properly with the climate integration and this change should be more default.

I haven't any climate device so I can very well have made some basic mistakes with the climate integration.

What are your thoughts here?

KartoffelToby commented:
@Kane610
My initial idea was to add a deCONZ Intigration Setting to toggle all deCONZ climate devices in remapping mode for auto and heat, so if your device like (the euronics spriti or i read about a Tuya/Moes TRV (TS0601) seems to have the same problems) sould used auto as heat remapping. Or some configuration.yaml setting to specify a device type and than remap some HVAC modes.

i mean, sure this change could be in deCONZ self. but at this time deCONZ haven't any climate specific handling and have different devices handlings

frenck commented:
> i mean, sure this change could be in deCONZ self. but at this time deCONZ haven't any climate specific handling 

That sounds incorrect, as the deCONZ rest plugin does have specific handling, even for the SPZB0001. The point here is that deCONZ handles device specifics in general. Adding another layer of quirk handling in Home Assistant is something we should not do. For example, if deCONZ starts fixing/adding quirks, then these types of modifications on our end may become problematic.

That said, in general, its not Home Assistant's job to fix things upstream. And, as indicated by Kane610, this might as well be a general implementation issue (which is in that case not fixed either). Either way, I feel like this implementation is incorrect.

KartoffelToby commented:
thanks @frenck unfortunately that's probably true, although I think remapping via config would be a nice thing.

anyway, this is a great project and i looking forward deCONZ and HA are great together.

I'll try to find some time to write a PR for deCONZ to match the Sprit device with the preset syntax for boost mode.

sadly i have only the sprit devices with are useless until a fix. But soon i will have a essentials Smart Home to test with deCONZ and HA

MartinHjelmare commented:
Should we close this PR?

KartoffelToby commented:
think we need a other oppinion

MartinHjelmare commented:
I don't know details about this integration but generally we want device specific details as far upstream as possible. Preferably we should receive a uniform interface from the library in a way that handles the device variability and let's us expose this as good as possible to the user.
